ibtk/src/math/PETScMatUtilities.cpp

```diff
@@ -806,10 +618,460 @@ void PETScMatUtilities::constructPatchLevelSCInterpOp(Mat& mat,
+
+        for (int axis = 0; axis < NDIM; ++axis)
+        {
+            IntVector<NDIM> offset = 0;
+            offset(axis) = 1;
+
+            int data_offset = 0;
+            for (int side = 0; side < axis; ++side)
+            {
+                int side_offset = depth;
+                for (unsigned d = 0; d < NDIM; ++d) side_offset *= coarse_num_cells[side](d);
+                data_offset += side_offset;
+            }
+
+            for (Box<NDIM>::Iterator b(SideGeometry<NDIM>::toSideBox(fine_patch_box, axis)); b; b++)
+            {
+                const CellIndex<NDIM>& i = b();
+                const SideIndex<NDIM> i_s(i, axis, SideIndex<NDIM>::Lower);
+                bool on_proc_fine_loc = true;
+                for (unsigned d = 0; d < depth; ++d)
+                {
+                    local_row[d] = (*fine_dof_data)(i_s, d);
+
+                    on_proc_fine_loc = on_proc_fine_loc && local_row[d] >= i_fine_lower && local_row[d] < i_fine_upper;
+
+                    local_row[d] -= i_fine_lower;
+                }
+                if (!on_proc_fine_loc) continue;
+
+                const CellIndex<NDIM> I = IndexUtilities::coarsen(i, fine_coarse_ratio);
+                const CellIndex<NDIM>& I_L = I;
+                const CellIndex<NDIM> I_U = I_L + offset;
+                for (unsigned d = 0; d < depth; ++d)
+                {
+                    samrai_petsc_map[d] = IndexUtilities::getIntegerMapping(
+                        I_L, coarse_domain_lower, coarse_num_cells[axis], d, coarse_ao_offset + data_offset);
+                    samrai_petsc_map[depth + d] = IndexUtilities::getIntegerMapping(
+                        I_U, coarse_domain_lower, coarse_num_cells[axis], d, coarse_ao_offset + data_offset);
+                }
+                AOApplicationToPetsc(coarse_level_ao, 2 * depth, &samrai_petsc_map[0]);
+
+                for (unsigned d = 0; d < depth; ++d)
+                {
+                    if (samrai_petsc_map[d] >= j_coarse_lower && samrai_petsc_map[d] < j_coarse_upper)
+                    {
+                        d_nnz[local_row[d]] = 1;
+                    }
+                    else
+                    {
+                        o_nnz[local_row[d]] = 1;
+                    }
+
+                    if (samrai_petsc_map[depth + d] >= j_coarse_lower && samrai_petsc_map[depth + d] < j_coarse_upper)
+                    {
+                        d_nnz[local_row[d]] += 1;
+                    }
+                    else
+                    {
+                        o_nnz[local_row[d]] += 1;
+                    }
+                }
+            }
+        }
+    }
+
+    // Create an empty matrix.
+    ierr = MatCreateAIJ(PETSC_COMM_WORLD, m_local, n_local, PETSC_DETERMINE, PETSC_DETERMINE, PETSC_DEFAULT, &d_nnz[0],
+                        PETSC_DEFAULT, &o_nnz[0], &mat);
+    IBTK_CHKERRQ(ierr);
+
+// Set some general matrix options.
+#if !defined(NDEBUG)
```

Comment:
I will. But when I read these lines, it somehow assures me that whatever nonzero spacing I have figured out, I am confident about it . It makes me happy to see these files, somehow :-)